### DISCUSSION TITLE
How to access Prisma client in both next.js router and express server

### DISCUSSION THREAD

**Discussion by withkynam** [prisma/prisma#22740] [Q&A]
### Question

In my full stack project, I have 2 folders client and backend. My frontend folder is a next.js project (which might contain serverless logic and endpoints) and my backend folder is a standalone persistent express server. 

I'm trying to get both to have access to my supabase using the prisma ORM. However, only 1 prisma.schema can exist and therefore only 1 prisma client should exists. 

I have seen prisma client being generated for either standalone server or within next.js, but not both. Is it possible?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @knamnguyen 👋 

Thank you for raising this question.

You can instantiate PrismaClient outside the client and backend folders and then use them in both of these folders. I think that should work. To achieve this, you need to structure your project in a way that allows both the backend and frontend to share the same Prisma client. See figure below

project-root
├── client
│   ├── next.js files
├── backend
│   ├── express server files
├── prisma
│   ├── schema.prisma
│   

By organizing your project in this way, you can maintain a single Prisma client that is used by both your backend and frontend,

**Reply by withkynam** [reply]
@ludralph Hi, thank you so much for your help. I tried to implement a monorepo structure as you suggested, but I could not manage to get it to work when deploying to vercel. My project's structure is as you mentioned, with a root folder containing the nextjs client folder, an express backend folder, and a prisma folder. In the root folder, I init a new typescript project and npm install the prisma client:

- In the root folder /root/db.ts, I initialize & export the prisma client in db.ts
- In the next.js client folder, I import the prisma client at root/client/src/app/server/lib
/db.ts:

```
import { prisma } from "../../../../../db";

export const prismaNextClient = prisma;
```

- In my next.js client, I'm importing again the prismaNextClient whereever I need to interact with the database, for example:

```
import { prismaNextClient } from "./../../server/lib/db";
...
const createdUser = await prismaNextClient.user.create({
    data: user,
  });

```

When I npm run dev for the nextjs folder locally, everything works great. However, when I deploy to vercel, I encounter the following error:

```
Failed to compile.
--
23:38:37.952 |  
23:38:37.952 | ../db.ts
23:38:37.952 | Module not found: Can't resolve '@prisma/client'
23:38:37.953 |  
23:38:37.953 | https://nextjs.org/docs/messages/module-not-found
23:38:37.953 |  
23:38:37.953 | Import trace for requested module:
23:38:37.953 | ./src/app/server/lib/db.ts
23:38:37.953 | ./src/app/api/webhook/route.ts
```

My vercel project is referencing the entire monorepo on github (the root folder containing everything), and I'm specifying the nextjs entry point as root/client (otherwise vercel wouldn't compile it as a nextjs project). There is a setting which says "Include source files outside of the Root Directory in the" which I have already checked but it didn't seem to do anything.

I noticed that my problem is similar to https://github.com/prisma/prisma/discussions/12350 - and I'm trying to follow the monorepo solution for easier future maintenance. But I'm not sure how to put it in production yet. Could you help?

**Reply by Harbinder04** [reply]
I am facing a similar issue. Have you found a solution?
My project scenario: [EduMastery Project Url](https://github.com/Harbinder04/Edu-mastery-app)
I am encountering the following error in my client folder (Next.js), while it works fine with the other two Express services.
**Error:** 

> PrismaClientInitializationError: Prisma Client could not locate the Query Engine for runtime "debian-openssl-3.0.x". This is likely caused by a bundler that has not copied "libquery_engine-debian-openssl-3.0.x.so.node" next to the resulting bundle. Ensure that "libquery_engine-debian-openssl-3.0.x.so.node" has been copied next to the bundle or in "node_modules/.prisma/client

I am not very experienced, so could you please suggest a solution?

**Reply by Harbinder04** [reply]
For the future reference, I got the solution here: 
[https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-monorepo](https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-monorepo)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
